**Exercise 1.21**

*Give a try of smallest-division.*


`Smallest-divison` at the begining of this section. 

In [2]:
(define (smallest-divisor n) (find-divisor n 2))
(define (find-divisor n test-divisor)
(cond ((> (square test-divisor) n) n)
      ((divides? test-divisor n) test-divisor)
      (else (find-divisor n (+ test-divisor 1)))))
(define (divides? a b) (= (remainder b a) 0))

divides?

In [7]:
(smallest-divisor 199)

199

In [8]:
(smallest-divisor 1999)

1999

In [9]:
(smallest-divisor 19999)

7

**Exercise 1.22**

*Introduced a print `display` procedure. Introduced a `runtime` procedure. Do the ***Order of Growth*** tests.* 

Done this before used `time` function at a online Scheme env which `runtime` is not available. Syntax `(time procedure)`.

The order of growth $\sqrt{n}$.

In [2]:
;smallest-divisor
(define (smallest-divisor n)
  (find-divisor n 2))
(define (find-divisor n test-divisor)
  (cond ((> (square test-divisor) n) n)
        ((divides? test-divisor n) test-divisor)
        (else (find-divisor n (next test-divisor))))) ;1.23 Change from (+ test-divisor 1) to (next test-divisor) 
(define (divides? a b)
  (= (remainder b a) 0))

;prime?
(define (prime? n)
  (= n (smallest-divisor n)))

;next
(define (next n)
  (if (= 2 n) 3 (+ 2 n)))
;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(define (timed-prime-test n)
  (newline)
  (display n)
  (start-prime-test n (runtime)))
(define (start-prime-test n start-time)
  (if (prime? n)
      (report-prime (- (runtime) start-time))))
(define (report-prime elapsed-time)
  (display " *** ")
  (display elapsed-time))

report-prime

In [3]:
;search-for-primes
(define (search-for-primes start count)
  (if (> count 0) 
      (let ((x (if (divides? 2 start)
                   (+ start 1)
                   start)))
        (if (prime? x)
            (begin
             (timed-prime-test x)
             (search-for-primes (+ x 2) (- count 1)))
            (search-for-primes (+ x 2) count)))))

search-for-primes

In [7]:
(search-for-primes 100000 3)


100003 *** 1.0000000000000009e-2
100019 *** 0.
100043 *** 0.

The primes AND runtimes are:

|Primes | 1.22 runtime(s) | 1.23 runtime | 1.24 runtime|
|-|-|-|-|
|1009 |0|0|0.03|
|1013 |0 |0|0.03|
|1019 |0 |0|0.04|
|100003|0|0|0.05|
|100019|0.01|0|0.05|
|100043|0.01|0|0.05|
|100000007|0.03|0.01|0.08|
|100000037|0.03|0.02|0.07|
|100000039|0.02|0.02|0.07|
|100000000003|0.84|0.55|0.12|
|100000000019|0.85|0.55|0.11|
|100000000057|0.84|0.52|0.11|




**Exercise 1.23**

*Although improved the performance, the Order of Growth is the same.$\sqrt{n}$*

Modified `smallest-divison` and a procedure `next` that returns 3 if its input is equal to 2 and otherwise returns its input plus 2. 

Change from `(+ test-divisor 1)` to `(next test-divisor)` and have a test.

In [7]:
;next
(define (next n)
  (if (= 2 n) 3 (+ 2 n)))

next

See result in above table. Improved some. According to the solutions website, it improved 2/3.

**Exercise 1.24**

*The Fermat method: Order of Growth* ***log(n).***

The result depends on the times each Fermat method runs the random test. If the times is small, the results would be too small to compare.Set `times` to 2000.

In [1]:
;fast-prime?
(define (expmod base exp m)
  (cond ((= exp 0) 1)
        ((even? exp)
         (remainder (square (expmod base (/ exp 2) m))
                    m))
        (else
         (remainder (* base (expmod base (- exp 1) m))
                    m))))        
(define (fermat-test n)
  (define (try-it a)
    (= (expmod a n n) a))
  (try-it (+ 1 (random (- n 1)))))
(define (fast-prime? n times)
  (cond ((= times 0) true)
        ((fermat-test n) (fast-prime? n (- times 1)))
        (else false)))

fast-prime?

In [2]:
;timed-prime-test
(define (timed-prime-test n)
  (newline)
  (display n)
  (start-prime-test n (runtime)))
(define (start-prime-test n start-time)
  (if (fast-prime? n 2000)
      (report-prime (- (runtime) start-time))))
(define (report-prime elapsed-time)
  (display " *** ")
  (display elapsed-time))

report-prime

In [3]:
;search-for-primes
(define (search-for-primes start count)
  (if (> count 0) 
      (let ((x (if (divides? 2 start)
                   (+ start 1)
                   start)))
        (if (fast-prime? x 500)
            (begin
             (timed-prime-test x)
             (search-for-primes (+ x 2) (- count 1)))
            (search-for-primes (+ x 2) count)))))
(define (divides? a b)
  (= (remainder b a) 0))

divides?

In [5]:
(search-for-primes 1000 3) 


1009 *** 2.9999999999999916e-2
1013 *** 4.0000000000000036e-2
1019 *** 4.0000000000000036e-2

**Exercise 1.25**

*A good example for calculate `mod` related problem.*

She is wrong. New expmod runs much slower. To compute $999999^{1000000}\mod1000000$, original one runs 0s, new expmod runs in 70 seconds.

Why?

In original `expmod`, operands are guaranteed to be smaller than mod, hence the product procedure will be faster.

In Alyssa version, operands will become extremely huge, and larger and larger in the process, so the product procedure is significantly slower.

For example:

```
(define (square m)  
   (display "square ")(display m)(newline) 
   (* m m)) 
  
 => (expmod 5 101 101) 
 square 5 
 square 24 
 square 71 
 square 92 
 square 1 
 square 1 
 5 
 => (remainder (fast-expt 5 101) 101) 
 square 5 
 square 25 
 square 625 
 square 390625 
 square 152587890625 
 square 23283064365386962890625 
 5 
```

In [1]:
(define (expmod-modified base exp m)
  (remainder (fast-expt base exp) m))
(define (fast-expt b n)
  (cond ((= n 0) 1)
        ((even? n) (square (fast-expt b (/ n 2))))
        (else (* b (fast-expt b (- n 1))))))
(define (even? n)
  (= (remainder n 2) 0))


(define (expmod base exp m)
(cond ((= exp 0) 1)
      ((even? exp)
        (remainder (square (expmod base (/ exp 2) m))
                  m))
      (else
        (remainder (* base (expmod base (- exp 1) m))
                  m))))    
(define (report-elapsed-time start-time) 
  (display " *** ")
  (display (- (runtime) start-time))
  (newline)) 

report-elapsed-time

Test the speed 

In [2]:
(define start-time (runtime)) 
(expmod 999999 1000000 1000000) 
(report-elapsed-time start-time)

 *** 0.


In [3]:
(define start-time (runtime)) 
(expmod-modified 999999 1000000 1000000) 
(report-elapsed-time start-time)
; 70.7

 *** 71.05


**Exercise 1.26**

*Implies that the slightest changes in code may lead to different time complexity.*

We know that there are applicative order and normal order. For this procedure,it is applicative order, so in the Louis's code, the operand of the `*`, will be evaluated twice.

In [ ]:
(define (expmod-new base exp m)
  (cond ((= exp 0) 1)
        ((even? exp)
         (remainder (* (expmod base (/ exp 2) m)
                       (expmod base (/ exp 2) m))
                    m))
        (else
         (remainder (* base (expmod base (- exp 1) m))
                    m))))

(define (expmod base exp m)
(cond ((= exp 0) 1)
      ((even? exp)
        (remainder (square (expmod base (/ exp 2) m))
                  m))
      (else
        (remainder (* base (expmod base (- exp 1) m))
                  m))))                        

**Exercise 1.27**

***Considering
an algorithm to be inadequate for the first reason but not for the second illustrates the
difference between mathematics and engineering.***

*In testing primality of very large numbers chosen at random, the chance of stumbling
upon a value that fools the Fermat test is less than the chance that cosmic radiation will
cause the computer to make an error in carrying out a “correct” algorithm.*

In [3]:
(define (expmod base exp m)
  (cond ((= exp 0) 1)
        ((even? exp)
         (remainder (square (expmod base (/ exp 2) m))
                    m))
        (else
         (remainder (* base (expmod base (- exp 1) m))
                    m))))        


(define (carmichael? n a)
  (cond ((>= a n) true)
        ((= (expmod a n n) a) (carmichael? n (+ a 1)))
        (else false)))

carmichael?

Carmichael numbers : The smallest few are 561, 1105, 1729, 2465, 2821, and 6601.

In [4]:
(carmichael? 561 1)

#t

In [5]:
(carmichael? 560 1)

#f

In [6]:
(carmichael? 1105 1)

#t

**Exercise 1.28**

***The Miller-Rabin test*** *that cannot
be fooled like the Fermat test.* 

In [5]:
(define (expmod base exp m)
  (cond ((= exp 0) 1)
        ((even? exp)
         (remainder (check-square (expmod base (/ exp 2) m) m)
                    m))
        (else
         (remainder (* base (expmod base (- exp 1) m))
                    m))))        
; a ^ (n - 1) modulo n == 1
; a ^ n modulo n == a;
;"nontrivial square root of 1 modulo n" : a number not equal to 1 or n - 1 whose square is equal to 1 modulo n
;   Hint: One convenient way to make expmod signal is to have it return 0.
(define (check-square a n)
  (define sqr (square a))
  (if (or (= a 1) (= a (- n 1)))
      sqr
      (if (= (remainder sqr n) 1)
          0
          sqr)))

(define (miller-rabin-test n)
  (define (try-it a)
    (= (expmod a (- n 1) n) 1))
  ;(try-it (+ 1 (random (- n 1)))))
(try-it 4))
(define (fast-prime? n times)
  (cond ((= times 0) true)
        ((miller-rabin-test n) (fast-prime? n (- times 1)))
        (else false)))

(define (fast-prime-count? n times false-count)
  (cond ((= times 0) false-count)
        ((miller-rabin-test n) (fast-prime-count? n (- times 1) false-count ))
        ((fast-prime-count? n (- times 1) (+ false-count 1)))))

fast-prime-count?

Try to analysis: a Miller–Rabin witness. (It is also possible to prove that if n is an odd number that is not prime, then, for at least half the numbers a < n, computing $a^{n - 1}$ in this way will reveal a nontrivial square root of 1 modulo n.) 

There is discuss about the wrong claim of miller-rabin witness in the book.https://stackoverflow.com/a/59834347

This one https://stackoverflow.com/a/74540926 explains well:
From [Cormen et al. 2009], p. 971:

Incorrect claim in SICP:

It is also possible to prove that if n is an odd number that is not prime, then, for at least half the numbers a < n, computing a^(n-1) in this way will reveal a nontrivial square root of 1 modulo n.

Corrected claim (equivalent to theorem 31.38 in [Cormen et al. 2009]):

It is also possible to prove that if n is an odd number that is not prime, then, for at least half the numbers a < n, computing a^(n-1) in this way will reveal a nontrivial square root of 1 modulo n **or a^(n-1) will not be equal to 1 module n.**


Test if n=21 is prime. The nontrivial square root of 1 modulo n only reveals in `8` and `13`.

|a| operand1 $a^{2}$|$a^{2\times2}$mod 21| operand2 $a^{5}$| $a^{5\times2}$mod 21| operand3 $a^{10}$| $a^{10\times2}$mod 21|
|-|-     |-      | -        | -| -| -|
| 2 | 4|16|11|16|16|4 |
| 3 | 9|18|12|18|18|9 |
| 4 | 16|4|16|4|4|16 |
| 5 | 4|16|17|16|16|4 |
| 6 | 15|15|6|15|15|15 |
| 8 | 1|1|8|1|1|1 |
| 9 | 18|9|18|9|9|18 |
| 10 | 16|4|19|4|4|16 |
| 11 | 16|4|2|4|4|16 |
| 12 | 18|9|3|9|9|18 |
| 13 | 1|1|13|1|1|1 |
| 14 | 7|7|14|7|7|7 |
| 15 | 15|15|15|15|15|15 |
| 16 | 4|16|4|16|16|4 |
| 17 | 16|4|5|4|4|16 |
| 18 | 9|18|9|18|18|9 |
| 19 | 4|16|10|16|16|4 |
| 20 | 1|1|20|1|1|1 |


In [9]:
(expmod 6 15 14)

6

In [6]:
(miller-rabin-test 13)

#t

In [10]:
(display (fast-prime? 561 10)) ;Carmichael numbers
(display (fast-prime? 1105 10)) ;Carmichael numbers
(display (fast-prime? 1729 10)) ;Carmichael numbers
(display (fast-prime? 1009 100)) ; prime

#f#f#f#t